### DISCUSSION TITLE
How to create tRPC client with a dynamic API URL in different environments

### DISCUSSION THREAD

**Discussion by kirgene** [trpc/trpc#4370] [Q&A]
I have a next.js application built as a Docker image. I want to use the same docker image in different environments.
I want to provide API URL for tRPC client (createTRPCNext) from process.env.
What's the proper way to do this?

**Comment by c-ehrlich** [comment]
basic example:
```ts
const trpcUrl = process.env.NEXT_PUBLIC_TRPC_URL ?? "http://localhost:3000";

export const api = createTRPCNext<AppRouter>({
  config() {
    return {
      transformer: superjson,
      links: [
        // maybe other links also
        httpBatchLink({
          url: `${trpcUrl}/api/trpc`,
        }),
      ],
    };
  },
});
```

note that you probably don't need an api url outside of SSR.

here's an example from create-t3-app:

```ts
const getBaseUrl = () => {
  if (typeof window !== "undefined") return ""; // browser should use relative url
  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`; // SSR should use vercel url
  return `http://localhost:${process.env.PORT ?? 3000}`; // dev SSR should use localhost
};
```

**Comment by kirgene** [comment]
Unfortunately, this won't work if I want to use a full custom URL (not just a base URL).
So the following code will fail on a client-side with `url is undefined` error (notice I removed `/api/trpc` from url):

```ts
const trpcUrl = process.env.NEXT_PUBLIC_TRPC_URL;

export const api = createTRPCNext<AppRouter>({
  config() {
    return {
      transformer: superjson,
      links: [
        // maybe other links also
        httpBatchLink({
          url: trpcUrl,
        }),
      ],
    };
  },
});
```

I want to be able to define a dynamic URL from env (for instance in `getInitialProps`) and then get generated tRPC client JS with that URL.
Something like this:
https://github.com/vercel/next.js/discussions/22243#discussioncomment-382394
